Add `FinalizationActor`

jordanschalm opened the pull request:
This PR extends https://github.com/onflow/flow-go/pull/4145:
- ~Removes all instances of `FinalizedHeaderSnapshot` and replace with `state.Final().Head()`~
  - As per discussion, we will instead keep `FinalizedHeaderSnapshot`, add a note that its primary use compared to `state.Final().Head()` is to avoid cache contention. I have replaced the existing implementation with one based on `FinalizationActor`.
- Implements `FinalizationActor`, as suggested in https://github.com/onflow/flow-go/pull/4145/files#r1160261030
  - In this implementation, we connect to HotStuff events and provide headers as `model.Block` -- we could connect to the protocol state events and use `flow.Header` instead (or do both)
  - Hooks up a `FinalizationActor` to `consensus/compliance.Engine` as an example usage

⚠️  The old `FinalizedSnapshot` was added to the RPC engine on `master`, making this non-mergeable without addressing those changes, which is done in https://github.com/onflow/flow-go/pull/4313.

durkmurder reviewed (commented):
I still think that we shouldn't deprecate `FinalizedHeaderCache`. I think it should be a standalone component which uses `FinalizationActor`. I believe the main benefit is it helps to have it in modules which aren't components. For instance, `API` backend will return finalized block for each request, it's better to serve it from `FinalizedHeaderCache` instead of using `state.Final().Head()`(which also sends wrong message to developers). Alternatively API can be a `component`. To summarize: I don't see any harm in keeping `FinalizedHeaderCache` and using it in some places.

jordanschalm commented:
> For instance, API backend will return finalized block for each request, it's better to serve it from FinalizedHeaderCache instead of using state.Final().Head()(which also sends wrong message to developers)

At the risk of bikeshedding... 👇 😬 

I'm still not quite on the same page as you and Alex regarding sending the wrong message to developers. I'm OK with keeping `FinalizedHeaderCache` in the way you suggested or removing it -- but I would like to understand the perspective which makes you feel that using `state.Final()` is sending the wrong message. From our earlier conversations, I agree there  might (or might not) be performance concerns with serving all finalized headers from one global cache, which would be a separate argument for something like `FinalizedHeaderCache`, but let's set that aside for now.

The message we don't want to send, from my perspective, is "please don't poll another component, or the database, in order to observe and process state changes -- instead communicate state changes asynchronously using events".

But making a query upon receiving a message in order to process it is different from polling vs pushing to respond to state changes.

For example, in the sync engine we do something like:
```go
func processSyncRequest(pingMsg) {
    finalizedHeader := getFinalizedHeader()
    if pingMsg.Height > finalizedHeader.Height {
        // maybe queue up some heights to request
    }
    conduit.Unicast(syncResponse{Height: finalizedHeader.Height})
}
```
Upon receiving a ping message, we retrieve the latest finalized block, and use its height to determine how to process the message. To me it makes little difference whether `getFinalizedHeader` uses `FinalizedHeaderCache` or the protocol state. We aren't polling the state for the latest finalized block, then processing based on changes we observe -- we are querying it upon receiving a separate message, in order to process the message. So long as `getFinalizedHeader` is fast, correct and consistent, it doesn't really matter whether it reads a cache that was set after a database transaction, or a cache that was set... oh right, after the same database transaction, with the same value, but asynchronously a few microseconds later.

I can imagine a perspective which says using `state.Final()` is bad because our component didn't get the finalized header from an event. Taken to the extreme: all components must receive all their inputs asynchronously via events and have no shared stateful dependencies (otherwise they could be used to communicate without using events). 

For the latest finalized block, it is fairly feasible for every component to follow this rule by receiving `OnBlockFinalized` events and maintaining a local copy -- it's just one value and it changes in a pretty straightforward way. But what about, for example, message processing where we query for a particular block by ID? Should the `ingestion` engine process `BlockIncorporated` events, and maintain a local mapping of block IDs to blocks, in order to validate the reference blocks of collections? This follows the same pattern, but definitely seems to cross a line into being unfeasible (or at least inadvisable).

durkmurder commented on engine/consensus/compliance/engine.go:
```diff
@@ -51,6 +49,7 @@ func NewEngine(
 	log zerolog.Logger,
 	me module.Local,
 	core *Core,
+	actor *events.FinalizationActor,
```
I am not sure about passing actor as argument, suddenly we are injecting dependency instead of using pub-sub. 
How about embedding `actor` in structure and subscribing externally on `FinalizationActor.OnBlockFinalized`. This way we will get benefits from using `FinalizationActor` but still preserve a pub-sub style subscription to avoid hard dependency.

jordanschalm replied:
Changed the construction pattern in https://github.com/onflow/flow-go/pull/4187/commits/b8d38e6bec80ced298aeecceff262df23ba350ee

durkmurder reviewed (approved):
Very nice, solid refactoring, looking forward seeing it merged and adopted in other engines.

jordanschalm commented on utils/unittest/unittest.go:
```diff
@@ -122,15 +122,13 @@ func SkipBenchmarkUnless(b *testing.B, reason SkipBenchmarkReason, message strin
-func ExpectPanic(expectedMsg string, t *testing.T) {
+// ExpectPanic fails the test if the calling thread did not panic.
+// Must be invoked as a deferred function.
+func ExpectPanic(t *testing.T) {
```
Good point - removed in https://github.com/onflow/flow-go/pull/4187/commits/45c076dedd0f305e00b2059d53e116ce66fb9023

jordanschalm commented on cmd/node_builder.go:
```diff
@@ -264,6 +264,8 @@ type NodeConfig struct {
 	RootChainID flow.ChainID
 	SporkID     flow.Identifier
+	// cached finalized block for use in bootstrapping
+	FinalizedHeader *flow.Header
```
https://github.com/onflow/flow-go/pull/4187/commits/a6f6fd591aa6d776b50ec47c6d83c79461a9eb20

AlexHentschel commented on module/events/finalization_actor.go:
```diff
@@ -0,0 +1,70 @@
+	}
+}
+
+// OnBlockFinalized receives block finalization events. It updates the newest finalized
+// block tracker and notifies the worker thread.
+func (actor *FinalizationActor) OnBlockFinalized(block *model.Block) {
```
I have a strong preference to using consistent method names for all the consumers. At the moment, the interface specifies https://github.com/onflow/flow-go/blob/a6f6fd591aa6d776b50ec47c6d83c79461a9eb20/consensus/hotstuff/consumer.go#L33

At the moment, we use anonymous methods for `OnFinalizedBlock` in various places instead of implementing the interface. Therefore, the current code compiles. However, I consider the anonymous methods approach an anti-pattern that I would really like to move away from. Quite frequently, I am wondering which components consume a specific event. But due to the anonymous methods, I can't just search for `OnFinalizedBlock` in the code base. Instead, I need to start at the distributor, figure out that you are routing the events through an anonymous method [here](https://github.com/onflow/flow-go/blob/a6f6fd591aa6d776b50ec47c6d83c79461a9eb20/module/events/finalization_actor.go#L11-L13). So in summary, I realize that the event is processed by the function `handleFinalizedBlock` in the compliance.engine while some other component processes the same event with a function `bar` and another one with a function `foo`. Am I able to describe the challenge?
Overall, I am just missing a clear naming convention, and without a convention there tend to be lots of different inconsistent patters making it super hard to find things.

Generally, I strongly prefer that all consuming functions for an event are named exactly as the interface -- generally the consumer just implements the interface.

jordanschalm replied:
I've made a few changes:
- In most cases, implement the interface directly
- When subscribing just one function (eg. `AddOnFinalizedBlockConsumer`), use the same function name

AlexHentschel commented on engine/collection/compliance/engine.go:
```diff
@@ -166,23 +154,12 @@ func (e *Engine) OnSyncedClusterBlock(syncedBlock flow.Slashable[*messages.Clust
-			}
-			e.core.ProcessFinalizedBlock(finalHeader)
-		}
+func (e *Engine) handleFinalizedBlock(block *model.Block) error {
```
Following up on my [previous comment](), I would strongly prefer a naming very close to the interface-specified consumer method  [`OnFinalizedBlock`](https://github.com/onflow/flow-go/blob/a6f6fd591aa6d776b50ec47c6d83c79461a9eb20/consensus/hotstuff/consumer.go#L28-L33)



```suggestion
// onFinalizedBlock informs compliance.Core about finalization of the respective block.
// The input to this callback is treated as trusted. This method should be executed on
// `OnFinalizedBlock` notifications from the node-internal consensus instance.
// No errors expected durning normal operations.
func (e *Engine) onFinalizedBlock(block *model.Block) error {
```

AlexHentschel commented on engine/consensus/compliance/engine.go:
```diff
@@ -175,23 +162,11 @@ func (e *Engine) OnSyncedBlocks(blocks flow.Slashable[[]*messages.BlockProposal]
-			}
-			e.core.ProcessFinalizedBlock(finalHeader)
-		}
+func (e *Engine) handleFinalizedBlock(block *model.Block) error {
```
```suggestion
// onFinalizedBlock informs compliance.Core about finalization of the respective block.
// The input to this callback is treated as trusted. This method should be executed on
// `OnFinalizedBlock` notifications from the node-internal consensus instance.
// No errors expected durning normal operations.
func (e *Engine) onFinalizedBlock(block *model.Block) error {
```

AlexHentschel commented on follower/follower_builder.go:
```diff
@@ -275,31 +274,17 @@ func (builder *FollowerServiceBuilder) buildFollowerEngine() *FollowerServiceBui
 			node.Metrics.Engine,
 			node.Network,
 			node.Me,
+			node.State,
```
can't we subscribe the sync engine to the `builder.FinalizationDistributor` here?

AlexHentschel commented on module/events/finalization_actor.go:
```diff
@@ -0,0 +1,70 @@
+	"github.com/onflow/flow-go/module/irrecoverable"
+)
+
+// OnBlockFinalized is invoked when a new block is finalized. It is possible that
+// blocks will be skipped.
+type OnBlockFinalized func(block *model.Block) error
```
To me, the naming here seems problematic:
* `OnBlockFinalized` is different yet not easily distinguishable from the [`OnFinalizedBlock` notification](https://github.com/onflow/flow-go/blob/a6f6fd591aa6d776b50ec47c6d83c79461a9eb20/consensus/hotstuff/consumer.go#L28-L33)
* I would guess that the difference is too small to notice for most people (I didn't realize that both methods were different until quite a bit through this PR). 

I could imagine two different approaches:
* We name it identically to the [FinalizationConsumer's `OnFinalizedBlock` notification](https://github.com/onflow/flow-go/blob/a6f6fd591aa6d776b50ec47c6d83c79461a9eb20/consensus/hotstuff/consumer.go#L28-L33). Then, we have the benefits of being able to search for the method
* If we want to highlight that they are different (this method is allowed to block and notifications can be dropped), I would make the methods' names obviously different.
   * For example, we could name this `ProcessLatestFinalizedBlock` (not sure if the "latest" provides enough of a hint that blocks could be skipped)

Additionally, I have a soft preference to use an interface here over anonymous methods.

AlexHentschel commented on module/events/finalization_actor.go:
```diff
@@ -0,0 +1,70 @@
+// This worker is designed to respond to a newly finalized blocks on a best-effort basis,
+// meaning that it may skip blocks when finalization occurs more quickly.
+// CAUTION: This is suitable for use only when the handler can tolerate skipped blocks.
+type FinalizationActor struct {
```
I have a soft preference for the `FinalizationActor` to implement the `FinalizationConsumer` interface

jordanschalm replied:
Done in https://github.com/onflow/flow-go/pull/4187/commits/bff4663ef672e2e7eea3ef50f8e69dea09fe9e8f.

AlexHentschel commented on module/events/finalization_actor.go:
```diff
@@ -0,0 +1,70 @@
+// worker is the worker function exposed by the FinalizationActor. It should be
+// attached to a ComponentBuilder by the higher-level component using CreateWorker.
+// It processes each new finalized block by invoking the OnBlockFinalized callback.
+func (actor *FinalizationActor) worker(ctx irrecoverable.SignalerContext, ready component.ReadyFunc) {
```
can we rename this to `workerLogic` or similar? I feel that the naming of `ComponentWorker` is misleading, because the thing is _not_ a worker. It is the logic to be _executed by_ a worker ... but the latter is out of scope for this PR

jordanschalm replied:
The difference between the function describing a worker's logic and the worker itself, in the context of the `ComponentManager`, is essentially whether it's invoked as a goroutine. The "invoke as a goroutine" part is entirely managed within the scope of the `ComponentManager`; we don't really see types that have both `func workerLogic(...)` _and_ `func worker(...) { go workerLogic(...) }`, so I think the opportunity for confusion here is pretty small. 

I did change the name here, but FWIW I personally think the scheme of naming functions which will be invoked as workers `XYZWorker`, and the `ComponentWorker` type name, makes sense.

AlexHentschel commented on module/events/finalized_header_cache.go:
```diff
@@ -0,0 +1,64 @@
+	return nil
+}
+
+// NewFinalizedHeaderCache returns a new FinalizedHeaderCache subscribed to the given FinalizationDistributor,
+// and the ComponentWorker function to maintain the cache.
+// The caller MUST start the returned ComponentWorker in a goroutine to maintain the cache.
```
I am confused about the documentation here: 
> subscribed to the given FinalizationDistributor

Something like the following would make sense, but not sure it is correct:
```suggestion
// NewFinalizedHeaderCache returns a new FinalizedHeaderCache and the ComponentWorker logic.
// The caller MUST:
//   - subscribe the `FinalizedHeaderCache` (first return value) to the `FinalizationDistributor`
//     that is distributing the consensus logic's `OnFinalizedBlock` events
//   - start the returned ComponentWorker logic (second return value) in a goroutine to maintain the cache.
//
```

jordanschalm replied:
Yes the reference to "given FinalizationDistributor" was outdated - thank you 👍

AlexHentschel reviewed (approved):
Very clean, well-readable code 👏 Mostly stylistic comments. My two areas of concern are:
* I have a strong preference to consistently use  the same name for all consumers (e.g. everywhere where we consume block finalization events, we use the method name `OnFinalizedBlock`). When reviewing code and wondering which components consume a specific event, anonymous consumer methods frequently create significant obstacles.  (see [this comment](https://github.com/onflow/flow-go/pull/4187#discussion_r1188027346) for more details)
* I feel that by moving to https://github.com/onflow/flow-go/blob/a6f6fd591aa6d776b50ec47c6d83c79461a9eb20/engine/common/synchronization/engine.go#L287 in the sync engine we are reinforcing the poll model as opposed providing a a good example for event-driven design (see [this comment](https://github.com/onflow/flow-go/pull/4187#discussion_r1188027346) for more details)

jordanschalm commented on engine/common/synchronization/engine.go:
```diff
@@ -284,7 +284,10 @@ func (e *Engine) processAvailableResponses() {
 func (e *Engine) onSyncResponse(originID flow.Identifier, res *messages.SyncResponse) {
 	e.log.Debug().Str("origin_id", originID.String()).Msg("received sync response")
-	final := e.finalizedHeader.Get()
+	final, err := e.state.Final().Head()
```
I explained my reasoning in detail above, for why I disagree with this line of thinking: https://github.com/onflow/flow-go/pull/4187#issuecomment-1506112592. 

We have also discussed this last week and I suggested a rule of thumb for where to draw the line regarding when information should be communicated via an event:
> - An event should represent a significant state change in the system (eg. a new block has been finalized)
> - A component should subscribe to an event if it needs to perform business logic as a result of the state change

This rule of thumb:
- Forbids polling for state changes, because we want to encourage event-driven designs
- Discourages every component maintaining their own local copy of the subset of the global state they need to read, because doing so introduces a lot of complexity for little benefit in most cases

An example of following this rule is: a verification node subscribes to `OnBlockProcessable` and begins verifying a chunk it is assigned to in the newly processable block

An example of not following this rule is: an engine sometimes needs to know the latest value of variable `X`, **at times that are not related to when `X` changes**, so it subscribes to `OnXChanged` and updates a local copy of `X` every time it consumes the event. 

_Only_ updating a local cache of the latest finalized block, when consuming an `OnBlockFinalized` event, is for the same reason an example of not following this rule. As we've discussed, there are potentially performance reasons for having multiple local caches of the latest finalized block in particular, because it is commonly accessed. The business logic _for that cache_ must respond to `OnBlockFinalized` events in order to maintain the cache. The business logic _for the component using the cache_ may need to do this, or it may not. 

> * On the main path of the logic, we are polling the information from a central source (potentially many times the same value, before any update occurs)

There is a difference between polling and reading data. Polling is when we repeatedly query a value, _so that we can do something when that value changes_. Here, we read the data because we want to know it's current value -- we don't care when it changes. 

> * no event-driven pattern, no usage of `FinalizationActor`

We do not use the FinalizationActor in this engine because this engine _does not perform business logic in response to blocks being finalized_. This engine largely performs business logic in response to messages from other nodes (which are treated similarly to internal events). In this case, we are responding to the `SyncResponse` by enqueuing new heights which should be queried. When we respond to the `SyncResponse`, we need to know the current value for the latest finalized block (so we read it), but we don't care when it changes (so we don't subscribe to events describing when it changes).

> Would it very hard to switch to the FinalizationActor here?

It would not be hard to use the `FinalizationActor` here, but it would be a poor example for when it is appropriate to use.

jordanschalm commented:
@AlexHentschel I've addressed the feedback about using `FinalizationActor` in the sync engine as well, in a follow-up PR here: https://github.com/onflow/flow-go/pull/4339

jordanschalm commented:
bors merge
